sdk/src/vote/voting.rs

```diff
@@ -0,0 +1,314 @@
+use crate::vote::submission::VoteExecutionStatus;
+use crate::vote::{EpochVoteSubmissions, UniqueVote};
+use fil_actors_runtime::fvm_ipld_hamt::BytesKey;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::econ::TokenAmount;
+use primitives::{TCid, THamt};
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::collections::BTreeSet;
+
+/// Handle the epoch voting
+#[derive(PartialEq, Eq, Clone)]
+pub struct Voting<Vote> {
+    /// The epoch that the voting started
+    pub genesis_epoch: ChainEpoch,
+    /// How often the voting should be submitted by validators
+    pub submission_period: ChainEpoch,
+    /// The last voting epoch that was executed
+    pub last_voting_executed_epoch: ChainEpoch,
+    /// Contains the executable epochs that are ready to be executed, but has yet to be executed.
+    /// This usually happens when previous submission epoch has not executed, but the next submission
+    /// epoch is ready to be executed. Most of the time this should be empty, we are wrapping with
+    /// Option instead of empty BTreeSet just to save some storage space.
+    pub executable_epoch_queue: Option<BTreeSet<ChainEpoch>>,
+    pub epoch_vote_submissions: TCid<THamt<ChainEpoch, EpochVoteSubmissions<Vote>>>,
+}
+
+impl<Vote: UniqueVote + DeserializeOwned + Serialize> Voting<Vote> {
+    pub fn new<BS: Blockstore>(
+        store: &BS,
+        genesis_epoch: ChainEpoch,
+        period: ChainEpoch,
+    ) -> anyhow::Result<Voting<Vote>> {
+        Ok(Self {
+            genesis_epoch,
+            submission_period: period,
+            last_voting_executed_epoch: genesis_epoch,
+            executable_epoch_queue: None,
+            epoch_vote_submissions: TCid::new_hamt(store)?,
+        })
+    }
+
+    pub fn submit_vote<BS: Blockstore>(
+        &mut self,
+        store: &BS,
+        vote: Vote,
+        epoch: ChainEpoch,
+        submitter: Address,
+        submitter_weight: TokenAmount,
+        total_weight: TokenAmount,
+    ) -> anyhow::Result<Option<Vote>> {
+        // We are doing this manually because we have to modify `state` while processing the `hamt`.
+        // The current `self.epoch_vote_submissions.modify(...)` does not allow us to modify state in the
+        // function closure passed to modify.
+        let mut hamt = self.epoch_vote_submissions.load(store)?;
+
+        let epoch_key = BytesKey::from(epoch.to_be_bytes().as_slice());
```

Comment:
:warning: Careful with this one, it lead to a huge bug in the interoperability with Go that took me a while to debug. You should use this function to derive the representation of epoch keys https://github.com/consensus-shipyard/ipc-actors/blob/dc3eaee617cc6b4d3bd080c52ff82bb730b08cfd/sdk/src/lib.rs#L30